### PR TITLE
Fix shares_increment bug & Add Tests Module

### PR DESCRIPTION
**Pull request by louisowen6**
Found bugs in `shares_increment` logic:
1) Negative actions should be added first with self.shares_increment
2) Actions clipping should be performed AFTER not BEFORE the rounding actions by shares_increment

### CODE REVIEW

**Comment by louisowen6** [comment]
Add `patient` parameter to control whether we have to end the cycle when we're running out of cash or just don't buy anything until we got additional cash

File: finrl/env/env_stocktrading_cashpenalty.py [hunk 1]
```diff
@@ -296,68 +268,64 @@ def step(self, actions):
...
+                    actions = np.where(actions>0,0,actions)
+                    spend = 0
+                    costs = 0
+                else:
+                    # ... end the cycle and penalize
+                    return self.return_terminal(
+                        reason="CASH SHORTAGE",reward=self.get_reward()
+                    )
+            self.transaction_memory.append(actions)  # capture what the model's could do
             # verify we didn't do anything impossible here
```
Comment (spencerR1992) [inline finrl/env/env_stocktrading_cashpenalty.py:301]: Can we please remove this feature from this PR and plan to do it in a subsequent PR?

**Review by spencerR1992 — CHANGES_REQUESTED** [review]
Louis. 
Can you please remove the `self.patient` functionality from this pr

Also, can you please write a test to verify your share increment fix works? Please let me know if you need help on this.

File: finrl/env/env_stocktrading_cashpenalty.py (hunk 1 above)
Comment (louisowen6) [reply finrl/env/env_stocktrading_cashpenalty.py]: Sure, will create other PR for adding the `self.patient` functionality.

Comment (louisowen6) [reply finrl/env/env_stocktrading_cashpenalty.py]: Also adding a test for share_increment later. Thanks!

**Comment by louisowen6** [comment]
Added tests module for `shares_increment`

File: finrl/env/env_stocktrading_cashpenalty.py [hunk 2]
```diff
@@ -75,8 +67,8 @@ def __init__(
         self.dates = df[date_col_name].sort_values().unique()
         self.random_start = random_start
         self.discrete_actions = discrete_actions
+        self.patient = patient
```
Comment (spencerR1992) [inline finrl/env/env_stocktrading_cashpenalty.py:70]: Please remove

File: finrl/env/env_stocktrading_cashpenalty.py [hunk 3]
```diff
@@ -67,6 +58,7 @@ def __init__(
         cache_indicator_data=True,
         cash_penalty_proportion=0.1,
         random_start=True,
+        patient=False,
```
Comment (spencerR1992) [inline finrl/env/env_stocktrading_cashpenalty.py:61]: Please remove

File: tests/test_env/test_env_cashpenalty.py [hunk 4]
```diff
@@ -41,6 +41,33 @@ def test_zero_step(self):
...
+        self.assertEqual(holdings[1], 0.0)
+
+        hmax_mc = self.df[self.df['tic']=='AAPL'].head(2).iloc[-1]['close'] / aapl_first_close
+        actions = np.array([-0.12 * hmax_mc,0.0])
+        next_state, _, _, _ = env.step(actions)
+        holdings = next_state[1 : 1 + len(self.ticker_list)]
+        self.assertEqual(holdings[0], 10.0)
+        self.assertEqual(holdings[1], 0.0)
+
+
```
Comment (spencerR1992) [inline tests/test_env/test_env_cashpenalty.py:70]: Nice test!

**Pull request merged**